### ISSUE TITLE
disabling xworker.window

### ISSUE THREAD

**Issue by bugzpodder** [pyscript/polyscript#28]
I have not tested in detail but AFAIK xworker allows direct access to dom via xworker.window.document.

`postMessage`, `isWindowProxy` and `window` is automatically injected into xworker which is accessible to the underlying script.
This means that it is possible to perform XSS attacks on another user.

In pyodie it is possible to `unregisterJsModule(js)` (hoodmane was [involved](https://github.com/pyodide/pyodide/issues/959)) in order for the environment to not have access to js.document.
Is there a way to do the same for xworker (disabling window)?

This is not particularly significant for pyscript, until someone adds pyscript to an app that allows one user to write pyscript code for another user to run.  This is the same as, I can put some script on a page with dom access with no issues, however if I allow one user to write script tags for another user to run then it becomes an issue.

aside, I am not clear how isWindowProxy should be used within xworker..

**Comment by WebReflection** [comment]
P.S. `xworker.window.document` is the only way to deal with the DOM via workers and that is a **feature** that won't ever be removed from this project ... those are both proxied values out of *coincident* though, and *coincident* uses a robust unique identifier to ensure the channel can't be addressed or leak and again, the whole point of *coincident* is to drive 100% the main thread *from a worker* and it's not going anywhere as that's a feature and it has no XSS issues whatsoever because *you own the running code, nobody else does*. I hope this helps understanding where we are, what we offer, why, and why that's not going away at any time.

**Comment by bugzpodder** [comment]
Thanks for the clarification.  `you own the running code` is not the case for me.  

> P.S. xworker.window.document is the only way to deal with the DOM via workers and that is a feature that won't ever be removed from this project

I agree and I am definitely not asking for this feature to be removed.  I am saying that:

- not all workers need to deal with DOM.  pyscript certainly does but in some user scenarios where this is not needed.

- with `xworker.sync` we can whitelist arbitrary functions that I can actually control, assuming the code python code I am running might be out of my control.

**Comment by WebReflection** [comment]
OK, I think I've got it ... you are asking to avoid worker code to propagate through the main so that "*each user gets a worker and what happens in their worker stays in their worker*" ... if I am understanding this correctly, please let me know, so I cna think about a way to avoid that.

Arguably, for educational purposes, this might indeed be a requirement ... if I got it right ... but did I get it right?

**Comment by bugzpodder** [comment]
> each user gets a worker and what happens in their worker stays in their worker

yes exactly, thanks!  Again I am currently executing everything in main thread and might move to xworkers in the future so not urgent in anyway.

**Comment by WebReflection** [comment]
A few solutions:

  1. `unregisterJsModule("polyscript")` so that nothing is reachable from `polyscript` when code executes ... yet internals should work just fine
  2. in worker, `from polyscript import xworker; del xworker.window;` or others ... in the worker, `xworker` has everything configurable ... you can pre-pend that before any code that's going to be evaluated and that should work too ... remember? you own the code you run, so if it's you enabling code to evaluate for any user, you can at least guard that code won't expose what you don't want to expose

`postMessage` is needed in the class or nothing would ever work, reason it's not even configurable, exactly to prevent XSS from evil 3rd party scripts, but can still remove it from the `xworker` reference that's trashed into `polyscript` module as it's not used internally at all.

would any of this work? dare I say having an example that reproduces your issue would help the both of us too ...

**Comment by WebReflection** [comment]
btw, if the second solution works for you, remember to also `del polyscript.target` as that indirectly exposes a place on the dom to eventually inject stuff ... although luckily enough, `innerHTML` doesn't execute `<script>` tags anymore, but having any DOM reference means they can `element.ownerDocument.documentView` to access either the `document` or the `window` they run into.

**Comment by bugzpodder** [comment]
i definitely need xworker.sync but `del xworker.window` should do the trick, thanks

**Comment by WebReflection** [comment]
@bugzpodder btw, I have been thinking about this ... to be clear, if 2 users or two tabs of the same browser reach the same `main` thread, that's still client-side ... multiple users per worker doesn't mean they can interfere with each others, they all run code in their own machine, it's not that they can share anything across browsers or tabs or PCs ... I hope this is clear, which is also why I didn't get the XSS concern to start with ... the user, if it hurts the main, hurts only itself, you need a websocket or something else to let them hurt anyone else on the Web, but happy to learn more about what you are trying to do.

**Comment by bugzpodder** [comment]
I tried to delete polyfill.xworker.document only and it didn't work because xworker is a JSProxy or something.  anyway I found out window.document is needed for matplotlib and I couldn't really delete it without breaking anything.

```
which is also why I didn't get the XSS concern to start with ... the user, if it hurts the main, hurts only itself, you need a websocket or something else to let them hurt anyone else on the Web
```
@WebReflection This is actually not the case for me.  User A writes some python code -> this gets stored by the server in the database (think pyscript.com) -> User B logins into the site that happens to show case User A's script, user A's script is run (in a worker) without User B knowing exactly what's happening -> script proceeds to do some malicious things (eg steals login information).  

It's like github.com running some random javascript snippet from me when you are viewing this message.

**Comment by WebReflection** [comment]
Ok, I understand your case better now but that’s dangerous for plain JS too because you are in fact enabling untrusted code to your users.

mad you also noticed, if DOM access is dropped packages depending on it will break and from polyscript perspective it’s impossible to know if the executing Python comes from trusted or untrusted code.

To be honest, I took the CS online course at Harvard and their system had the same issue + codepen and other online tools suffer pretty much the same gotcha. Cookies and localStorage but also indexed DB can be abused **but** what you could do is to use an iframe with all the guards as main thread. That iframe won’t leak anything sensible and it can still be used as main.

In short, the platform has guards to prevent evil things and this library can still run behind those guards but it can’t prevent untrusted code to work.

**Comment by WebReflection** [comment]
See this post to know more how to disable sensible details from a user: https://developer.mozilla.org/en-US/docs/Web/Security/IFrame_credentialless

that dev can point at any page you like and your concerns should disappear.

**Comment by bugzpodder** [comment]
iframe is a good solution in many situations.  However on my page I am rendering multiple python script blocks and I cannot iframe them individually (else i have to load pyodide in each iframe).  I could iframe the entire thing that contains all python script blocks but I am not clear of the security implications especially since I have an authenticated user, so have to err on the safe side.

> As you also noticed, if DOM access is dropped packages depending on it will break and from polyscript perspective it’s impossible to know if the executing Python comes from trusted or untrusted code.

I think I found a way to switch to a different matplotlib backend that does not rely on js.document (have not tested it yet).
It would be great if polyscript.xworker.window can be disabled via a flag while keeping polyscript.xwoker.sync enabled.  I agree that polyscript can allow access to window.document by default, but if some flag can be passed that would disallow xworker.window (while keeping xwoker.sync) that might be acceptable solution.  Assuming `from polyscript import xworker; del xworker.window;` doesn't work.  I think I tried this on Thursday and didn't work, but I'll try it again at some point.

Anyway no big deal for now, I'll need to determine the feasibility of removing js.document on the main thread first.

**Comment by bugzpodder** [comment]
ok was able to use your linkedom and create a fake document node in order to satisfy matplotlib..

**Comment by WebReflection** [comment]
so ... we're good?

> I could iframe the entire thing that contains all python script blocks but I am not clear of the security implications especially since I have an authenticated user, so have to err on the safe side.

CodePen and many others do that for logged in users ... it's not straight forward, I suppose, but there is surely a way.

**Comment by bugzpodder** [comment]
> so ... we're good?

ok looks like `del polyscript.xworker` and `del polyscript.xworker.window` both worked.
